### PR TITLE
Switch from `watchgod` to `watchfiles`

### PR DESCRIPTION
**Pull request by samuelcolvin**
See https://watchfiles.helpmanual.io/migrating/.

TL;DR; I've renamed `watchgod` to `watchfiles` and rewritten it to use the rust "notify" crate for all the heavy lifting. This PR is a WIP to see if you're willing to move ahead with a change like this.

Still TODO/to answer:
* I've changed the logic of `BaseReload` to try and make `StatReload` and watchfiles share an interface, are you happy with this?
* tests are a fair bit more complicated, are you happy with how I'm implemented/changed them?
* ~~I don't see an obvious way to deal with `--reload-include` and `--reload-exclude` being paths rather than extensions. Are you happy to drop this functionality? It's much less important now there's little overhead to scanning big directories, if not, we can implement something but it'll be complicated.~~ I think I've dealt with this a287d5c

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Would be great if someone could kick off tests?

**Comment by samuelcolvin** [comment]
From https://github.com/encode/uvicorn/settings/actions the below option might save you a lot of time kicking off tests and would be helpful for new contributors.

image

**Comment by Kludex** [comment]
Ah... I didn't know it was configurable. I'm going ask @tomchristie to enable that over encode. Thanks 🙏

Ping @florimondmanca as you asked about this the other day.

**Comment by samuelcolvin** [comment]
> Ah... I didn't know it was configurable. I'm going ask @tomchristie to enable that over encode. Thanks 🙏

Ye, took me some time to find it, it's really helpful.

**Comment by samuelcolvin** [comment]
I would love feedback on this, but please don't merge. I'd like to make some more changes once https://github.com/samuelcolvin/watchfiles/pull/113 is merged and released (mostly making sure tests don't wait forever if no changes are detected).

**Comment by Kludex** [comment]
I think I'll not have time to review this today, but my initial thoughts are to first deprecate watchgod, and not to remove it directly. As on the user's point of view those are different packages, and it would be a breaking change (even if that package is not supposed to be used in production on uvicorn...).

But... I think it's fine to change all documentation and replace the standard extra requirements. About the tests, I'll need to check.

**Comment by samuelcolvin** [comment]
Makes sense, should be possible to add back the watchgod class, then choose between the three.

**Comment by samuelcolvin** [comment]
AKAIK this is ready to review.

File: .gitignore
Comment (Kludex) [inline .gitignore]: We reject any addition to the `.gitignore` that doesn't come from the repo build or specific tool in here.

Ref.: https://github.com/encode/uvicorn/pull/1001

```suggestion
```

File: setup.cfg
Comment (Kludex) [inline setup.cfg:25]:
```suggestion
    uvicorn/supervisors/watchfilesreload.py,
    uvicorn/supervisors/watchgodreload.py, 
```

File: setup.cfg
Comment (Kludex) [inline setup.cfg]:
```suggestion
known_third_party = click,does_not_exist,gunicorn,h11,httptools,pytest,requests,setuptools,urllib3,uvloop,watchgod,watchfiles,websockets,wsproto,yaml
```

File: uvicorn/supervisors/watchgodreload.py
Comment (Kludex) [inline uvicorn/supervisors/watchgodreload.py]: Can we have the message below the `WatchGodReload` so the message gets printed on import, and not on instantiation?
```suggestion
            "to watchfiles (`pip install watchfiles`).",
```

File: uvicorn/supervisors/basereload.py
Comment (Kludex) [inline uvicorn/supervisors/basereload.py:112]: I know we had this before, but do you know when do we have a `ValueError`?

File: uvicorn/supervisors/watchgodreload.py
Comment (Kludex) [inline uvicorn/supervisors/watchgodreload.py:156]: Isn't `change` already a set? Is it possible to have duplicated paths here?
```suggestion
                return [Path(c[1]) for c in change]
```

File: uvicorn/supervisors/watchgodreload.py
Comment (Kludex) [inline uvicorn/supervisors/watchgodreload.py:150]: The `self.pause` is not being called here, is it expected?

File: uvicorn/supervisors/watchfilesreload.py
Comment (Kludex) [inline uvicorn/supervisors/watchfilesreload.py:84]: The `self.pause()` is not being called here. Is it expected?

**Review by Kludex — COMMENTED** [review]
> I've changed the logic of BaseReload to try and make StatReload and watchfiles share an interface, are you happy with this?

I'm fine with the changes on the `BaseReload` class. It's easier to understand the logic of `run()`. Thanks. 👍

I don't care about the change on the `reloader_name` for each reloader class. Do you @euri10 ?

I've made an initial review. I didn't check the tests yet.

**Comment by Kludex** [comment]
Thanks for the PR 🙇

File: uvicorn/supervisors/basereload.py
Comment (samuelcolvin) [reply uvicorn/supervisors/basereload.py:112]: Lots! Whenever `path` is not inside `cwd`.

File: uvicorn/supervisors/watchfilesreload.py
Comment (samuelcolvin) [reply uvicorn/supervisors/watchfilesreload.py:84]: This is correct.

```py
changes = next(self.watcher)
```

Waits for changes, but returns once a batch of changes have occurred - this is the whole point of watchfiles!

File: uvicorn/supervisors/watchgodreload.py
Comment (samuelcolvin) [reply uvicorn/supervisors/watchgodreload.py:150]: Good catch, fixed.

File: uvicorn/supervisors/watchgodreload.py
Comment (samuelcolvin) [reply uvicorn/supervisors/watchgodreload.py:156]: This is required, you often get a "modified" and "added" or "deleted" in the same batch.

**Review by samuelcolvin — COMMENTED** [review]
Suggestions added.

File: uvicorn/supervisors/watchgodreload.py
Comment (Kludex) [reply uvicorn/supervisors/watchgodreload.py]: Sorry, I meant some lines above.

```python
class WatchGodReload:
    warnings.warn(...)
```
So I can get a warning with those lines already:
```python
from uvicorn.supervisors.watchgodreload import WatchGodReload
```

Comment (samuelcolvin) [reply uvicorn/supervisors/watchgodreload.py]: pushed but tests will fail.

Make you could fix the warning handling in tests for this? I have no idea how you want to deal with this.

Honestly, I think the previous play where the warning was in init was better.

Comment (Kludex) [reply uvicorn/supervisors/watchgodreload.py]: I put it back on the `__init__`, sorry for the confusion here.

**Comment by Kludex** [comment]
Things failing here because of coverage. The warning is not an issue anymore.
```
uvicorn/supervisors/basereload.py                    65     11    83.08%   41, 44-54, 58, 61, 97, 106
uvicorn/supervisors/watchfilesreload.py              56      3    94.64%   36-38, 92
uvicorn/supervisors/watchgodreload.py               100      4    96.00%   112-120
```

**Comment by samuelcolvin** [comment]
@Kludex coverage should be fixed.

**Comment by Kludex** [comment]
ty, will check in the following day(s) 🙏

File: main.py
Comment (euri10) [inline main.py]: Should this be here ? doesnt seem it is used anywhere in the PR

File: tests/supervisors/test_reload.py
Comment (euri10) [inline tests/supervisors/test_reload.py:32]: save time :heart:

**Review by euri10 — APPROVED** [review]
looks very promising to me I tested it locally on a big project and one can feel the speed
I just have to minor comments and it needs to be refreshed to solve conflicts but its super cool to have, thanks

File: main.py
Comment (samuelcolvin) [reply main.py]: No idea, don't think I added this.

**Comment by samuelcolvin** [comment]
@Kludex sorry to hassle, why update on this?

**Comment by Kludex** [comment]
Update what?

**Comment by samuelcolvin** [comment]
> ty, will check in the following day(s) 🙏

You mentioned you'd check this.

**Comment by Kludex** [comment]
Ah! Sorry!

I'm going to check in some hours. 🙏

File: main.py
Comment (Kludex) [reply main.py]: 👀 🏃 💨

File: uvicorn/supervisors/watchfilesreload.py
Comment (Kludex) [inline uvicorn/supervisors/watchfilesreload.py]: The logger is not used in this file.
```suggestion
```

File: uvicorn/supervisors/watchfilesreload.py
Comment (Kludex) [inline uvicorn/supervisors/watchfilesreload.py]:
```suggestion
```

File: tests/supervisors/test_reload.py
Comment (Kludex) [inline tests/supervisors/test_reload.py]: Duplicated, I guess?

**Comment by Kludex** [comment]
@samuelcolvin Done! 👍 

Besides those 3 comments, I don't have more to say here. Thanks for the PR! 🙇

**Comment by Kludex** [comment]
I'll see today how's going to be our release schedule (i.e. if there's going to be a patch before merging this). Jfyk

**Review by Kludex — APPROVED** [review]

**Comment by Kludex** [comment]
Thanks @samuelcolvin ! 🙇 

Sorry for taking so long on the review.

**Comment by samuelcolvin** [comment]
No problem at all, thanks reviewing and all your work.

**Comment by sovanna** [comment]
I needed to install `rust` in order to install uvicorn _to make watchfiles works_.

`pip install uvicorn[standard]` didn't work for me anymore.

I had the issue below.

```
Collecting watchfiles
  Using cached watchfiles-0.15.0.tar.gz (36 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      Checking for Rust toolchain....

      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/

      [end of output]
```

_Using inside a Docker container (FROM pypy:3)_

**Comment by samuelcolvin** [comment]
Best to create an issue on watchfiles.

Currently we don't distribute binaries for pypy.

You can always skip installing watchfiles and uvicorn will fall back to polling.

**Comment by bropuffshroom** [comment]
> I needed to install `rust` in order to install uvicorn _to make watchfiles works_.
> 
> `pip install uvicorn[standard]` didn't work for me anymore.
> 
> I had the issue below.
> 
> ```
> Collecting watchfiles
>   Using cached watchfiles-0.15.0.tar.gz (36 kB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>   Preparing metadata (pyproject.toml) ... error
>   error: subprocess-exited-with-error
> 
>   × Preparing metadata (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [6 lines of output]
>       Checking for Rust toolchain....
> 
>       Cargo, the Rust package manager, is not installed or is not on PATH.
>       This package requires Rust and Cargo to compile extensions. Install it through
>       the system's package manager or via https://rustup.rs/
> 
>       [end of output]
> ```
> 
> _Using inside a Docker container (FROM pypy:3)_

Can confirm this. I ran fastapi in a docker and the reload just didn't work, but older version works.
This basically rendered --reload broken on uvicorn[standard] installation for some environment, without giving any warning or information whatsoever. I just spend a few hours to debug why the auto reload didn't work for me.

I personally think it's probably not a good idea to migrate to watchfiles for now just because watchgod is deprecated.

For my situation (I'm running a linux container on windows), I didn't run into any trouble installing uvicorn[standard] or watchfiles.

For anyone who ran into the same issue, as mentioned above, after the uvicorn[standard] installation, uninstall watchfiles and uvicorn will fall back to StatReload and the reload will be working.

**Comment by bropuffshroom** [comment]
> I needed to install `rust` in order to install uvicorn _to make watchfiles works_.
> 
> `pip install uvicorn[standard]` didn't work for me anymore.
> 
> I had the issue below.
> 
> ```
> Collecting watchfiles
>   Using cached watchfiles-0.15.0.tar.gz (36 kB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>   Preparing metadata (pyproject.toml) ... error
>   error: subprocess-exited-with-error
> 
>   × Preparing metadata (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [6 lines of output]
>       Checking for Rust toolchain....
> 
>       Cargo, the Rust package manager, is not installed or is not on PATH.
>       This package requires Rust and Cargo to compile extensions. Install it through
>       the system's package manager or via https://rustup.rs/
> 
>       [end of output]
> ```
> 
> _Using inside a Docker container (FROM pypy:3)_

Can confirm this. I ran fastapi in a docker and the reload just didn't work, but older version works.
This basically rendered --reload broken on uvicorn[standard] installation for some environment, without giving any warning or information whatsoever. 

In my very personal opinion I think it's probably not a good idea to migrate to watchfiles for now just because watchgod is deprecated, since I just spend a few hours to debug why the auto reload didn't work for me.

For my situation (I'm running a linux container on windows), I didn't run into any trouble installing uvicorn[standard] or watchfiles.

For anyone who ran into the same issue, as mentioned above, after the uvicorn[standard] installation, uninstall watchfiles and uvicorn will fall back to StatReload and the reload will be working.

**Comment by samuelcolvin** [comment]
Please create new issues rather than commenting on closed PRs. You're issue is not the same as this, both issues have been fixed.

Your issue looks to be the same as https://github.com/samuelcolvin/watchfiles/issues/169, you just need to set `WATCHFILES_FORCE_POLLING` and reloading will work.

There were many issues with watchgod, hence why I rewrote it to create watchfiles. As the author of both i can assure you watchfiles is much better in many ways.

**Comment by bropuffshroom** [comment]
> Please create new issues rather than commenting on closed PRs. You're issue is not the same as this, both issues have been fixed.
> 
> Your issue looks to be the same as [samuelcolvin/watchfiles#169](https://github.com/samuelcolvin/watchfiles/issues/169), you just need to set `WATCHFILES_FORCE_POLLING` and reloading will help.
> 
> There were many issues with watchgod, hence why I rewrote it to create watchfiles. As the author of both i can assure you watchfiles is much better in many ways.

Thanks for your work. Both are excellent libraries.

**Comment by dvglab** [comment]
> Please create new issues rather than commenting on closed PRs. You're issue is not the same as this, both issues have been fixed.
> 
> Your issue looks to be the same as [samuelcolvin/watchfiles#169](https://github.com/samuelcolvin/watchfiles/issues/169), you just need to set `WATCHFILES_FORCE_POLLING` and reloading will work.
> 
> There were many issues with watchgod, hence why I rewrote it to create watchfiles. As the author of both i can assure you watchfiles is much better in many ways.

Information about WATCHFILES_FORCE_POLLING is very important. I've lost a lot of hours before found it here. It seems to me it should be on uvicorn site somewhere in documentation. It was a real headache to realize why the reload function don't work on a docker for Windows. Thanks God I found it here.

**Comment by Kludex** [comment]
PR to improve docs are welcome.

**Pull request merged**
